examples/main/main.cpp

```diff
@@ -619,6 +620,11 @@ int main(int argc, char ** argv) {
         }
     }
 
+    if (!path_session.empty() && session_save_all) {
+        fprintf(stderr, "\n%s: saving final output to session file '%s'\n", __func__, path_session.c_str());
+        llama_save_session_file(ctx, path_session.c_str(), session_tokens.data(), session_tokens.size());
+    }
+
```

Comment:
This is the only real issue I see with this PR. Nothing else is really a problem.

There are two points where the program exits. Here, and in `sigint_handler`. If you're in interactive mode, the only way to exit is with Ctrl-C. So as is, for most people, they will end up quitting without their cache being written.

Now that I think about it though, it's probably not safe to call `llama_save_session_file` in `sigint_handler` because you could be in the middle of an evaluation.

~~That makes it tricky to just throw in there. Off the top of my head, we'd have to add another variable like `is_interacting` but named `is_evaluating` (which should both be `volatile sig_atomic_t` btw) that we set to true and false around our llama evals and and if someone tries to `Ctrl-C` while `is_evaluating` it would warn them that it may take a second to save or that they can hit `Ctrl-C` again to exit without saving and we'd set `params.interactive` to false, `n_remain` to 0, and `params.n_predict` to 0, which would trigger this exit path.~~ Edit: You'll still get stuck on input.

If you don't have experience with interrupts, threading, or race conditions, I'd skip that approach.

Next best solution would be write the session out during interactive mode every time before we prompt the user for input. Around [line 530](https://github.com/ggerganov/llama.cpp/blob/87d6f8c93379db372d06b570e316df69decbf018/examples/main/main.cpp#L530), which unfortunately, could be quite often. To be more optimal we could watch the token count and only save every 128 token evals (or some number that makes sense).

Once again you're stung by the whole interactivity part of `main`. I guess you could just state that it doesn't work in interactive or instructional mode in `gpt_print_usage` and in `README.md`, but that makes it hard to accept this pull request as is.